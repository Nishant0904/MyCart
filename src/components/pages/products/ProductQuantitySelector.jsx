import styled from 'styled-components'
import Button from '../../reusables/buttons/Button'
import { useState } from 'react'
import PropTypes from 'prop-types'

const ProductQuantitySelector = ({ product, handleAddToCart }) => {
    const [quantity, setQuantity] = useState(1)

    return (
        <StyledQuantitySelector data-testid='product-quantity-selector'>
            <QuantityInputContainer>
                <Button
                    onClick={handleDecrement}
                    data-testid='product-quantity__decrement-btn'
                >
                    â€“
                </Button>
                <input
                    type='number'
                    value={quantity}
                    onChange={handleInputChange}
                    onBlur={(e) => {
                        if (e.target.value === '') setQuantity(1)
                    }}
                    data-testid='product-quantity__number-input'
                />
                <Button
                    onClick={handleIncrement}
                    data-testid='product-quantity__increment-btn'
                >
                    +
                </Button>
            </QuantityInputContainer>
            <Button
                onClick={() => {
                    handleAddToCart(product, quantity)
                }}
                data-testid='product-quantity__add-to-cart-btn'
            >
                Add to Cart
            </Button>
        </StyledQuantitySelector>
    )

    function handleIncrement() {
        if (quantity < 10) setQuantity(quantity + 1)
    }

    function handleDecrement() {
        if (quantity > 1) setQuantity(quantity - 1)
    }

    function handleInputChange(e) {
        const newValue = e.target.value
        if (newValue <= 10 && newValue >= 0) {
            setQuantity(parseInt(newValue))
        } else if (newValue > 10 || newValue < 0) {
            setQuantity(1)
        } else {
            setQuantity('')
        }
    }
}

const StyledQuantitySelector = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;
`

const QuantityInputContainer = styled.div`
    height: 2rem;
    display: flex;
    gap: 2px;

    input {
        width: 2rem;
        text-align: center;
        border: none;
        &::-webkit-inner-spin-button,
        &::-webkit-outer-spin-button {
            -webkit-appearance: none;
            margin: 0;
        }
    }

    Button {
        display: flex;
        align-items: center;
        justify-content: center;
        padding: 8px;
        font-size: 1.2rem;
    }
`

ProductQuantitySelector.propTypes = {
    product: PropTypes.object.isRequired,
    handleAddToCart: PropTypes.func.isRequired,
}

export default ProductQuantitySelector
