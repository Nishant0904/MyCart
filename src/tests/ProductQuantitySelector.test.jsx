import { describe, it, expect, vi } from 'vitest'
import { render, screen } from '@testing-library/react'
import ProductQuantitySelector from '../components/pages/products/ProductQuantitySelector'
import { BrowserRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
}

describe('Product Quantity Selector test', () => {
    it('Product Quantity Selector renders correctly', () => {
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        expect(
            screen.getByTestId('product-quantity-selector'),
        ).toBeInTheDocument()
        expect(
            screen.getByTestId('product-quantity-selector'),
        ).toMatchSnapshot()
    })

    it('Increment button adds one to the input number value', async () => {
        const user = userEvent.setup()
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const incrementBtn = screen.getByTestId(
            'product-quantity__increment-btn',
        )
        const numberInput = screen.getByTestId('product-quantity__number-input')

        await user.click(incrementBtn)

        expect(numberInput.value).toBe('2')
    })

    it('Increment button can only add up to 10', async () => {
        const user = userEvent.setup()
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const incrementBtn = screen.getByTestId(
            'product-quantity__increment-btn',
        )
        const numberInput = screen.getByTestId('product-quantity__number-input')

        // click the button ten times, with a starting value of 1
        for (let i = 0; i < 10; i++) {
            await user.click(incrementBtn)
        }

        expect(numberInput.value).toBe('10')
    })

    it('Decrement button can not subtract below 1', async () => {
        const user = userEvent.setup()
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const decrementBtn = screen.getByTestId(
            'product-quantity__decrement-btn',
        )
        const numberInput = screen.getByTestId('product-quantity__number-input')

        await user.click(decrementBtn)

        expect(numberInput.value).toBe('1')
    })

    it('Decrement button can not subtract below 1', async () => {
        const user = userEvent.setup()
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={() => {}}
            />,
            { wrapper: BrowserRouter },
        )
        const incrementBtn = screen.getByTestId(
            'product-quantity__increment-btn',
        )
        const decrementBtn = screen.getByTestId(
            'product-quantity__decrement-btn',
        )
        const numberInput = screen.getByTestId('product-quantity__number-input')

        await user.dblClick(incrementBtn)
        await user.click(decrementBtn)

        expect(numberInput.value).toBe('2')
    })

    it('handleAddToCart is called on Add to Cart btn click', async () => {
        const user = userEvent.setup()
        const handleAddToCartMock = vi.fn()
        render(
            <ProductQuantitySelector
                product={mockProduct}
                handleAddToCart={handleAddToCartMock}
            />,
            { wrapper: BrowserRouter },
        )
        const addToCartBtn = screen.getByTestId('product-quantity__add-to-cart-btn')

        await user.click(addToCartBtn)

        expect(handleAddToCartMock).toHaveBeenCalled();
    })
})
