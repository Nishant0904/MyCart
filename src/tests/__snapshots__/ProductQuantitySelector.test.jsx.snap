// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Quantity Selector test > Product Quantity Selector renders correctly 1`] = `
<div
  class="sc-guDLey hGbZVM"
  data-testid="product-quantity-selector"
>
  <div
    class="sc-dmyCSP kIuZzL"
  >
    <button
      class="sc-beySPh gepDOA"
      data-testid="product-quantity__decrement-btn"
    >
      â€“
    </button>
    <input
      data-testid="product-quantity__number-input"
      type="number"
      value="1"
    />
    <button
      class="sc-beySPh gepDOA"
      data-testid="product-quantity__increment-btn"
    >
      +
    </button>
  </div>
  <button
    class="sc-beySPh gepDOA"
    data-testid="product-quantity__add-to-cart-btn"
  >
    Add to Cart
  </button>
</div>
`;
